File: day12/day12.cpp
#include <cassert>
#include <fstream>
#include <vector>
#include <fmt/format.h>

using namespace std;

struct Operand
{
	enum { REG, LIT } type;
	int value;

	int get(int regs[]) const noexcept
	{
		return type == REG ? regs[value] : value;
	}
};

istream& operator>>(istream& in, Operand& o)
{
	string s;
	if (in >> s)
	{
		if ("abcd"s.find(s[0]) != string::npos)
		{
			o.type = Operand::REG;
			o.value = s[0]-'a';
		}
		else
		{
			o.type = Operand::LIT;
			o.value = stoi(s);
		}
	}
	return in;
}

struct Instruction
{
	enum { CPY, INC, DEC, JNZ } opcode;
	Operand op1;
	Operand op2;
};

istream& operator>>(istream& in, Instruction& i)
{
	string s;
	in >> s;
	if (s == "cpy")
	{
		i.opcode = Instruction::CPY;
		in >> i.op1;
		in >> i.op2;
	}
	else if (s == "inc")
	{
		i.opcode = Instruction::INC;
		in >> i.op1;
	}
	else if (s == "dec")
	{
		i.opcode = Instruction::DEC;
		in >> i.op1;
	}
	else if (s == "jnz")
	{
		i.opcode = Instruction::JNZ;
		in >> i.op1;
		in >> i.op2;
	}
	else
	{
		in.setstate(ios::failbit);
	}
	return in;
}

struct Computer
{
	int registers[4];

	void reset() noexcept
	{
		fill(registers, registers + 4, 0);
	}

	int execute(const vector<Instruction>& program) noexcept
	{
		int ip = 0;
		while (0 <= ip && (size_t)ip < program.size())
		{
			auto& instr = program[ip];
			switch (instr.opcode)
			{
			case Instruction::CPY:
				assert(instr.op2.type == Operand::REG);
				registers[instr.op2.value] = instr.op1.get(registers);
				ip++;
				break;

			case Instruction::INC:
				assert(instr.op1.type == Operand::REG);
				registers[instr.op1.value]++;
				ip++;
				break;

			case Instruction::DEC:
				assert(instr.op1.type == Operand::REG);
				registers[instr.op1.value]--;
				ip++;
				break;

			case Instruction::JNZ:
				if (instr.op1.get(registers))
				{
					ip += instr.op2.get(registers);
				}
				else
				{
					ip++;
				}
				break;

			default:
				abort();
			}
		}
		return registers[0];
	}
};

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fmt::print(stderr, "Usage: {} <filename>\n", argv[0]);
		return 1;
	}

	ifstream input(argv[1]);
	if (!input)
	{
		fmt::print(stderr, "Cannot open {}\n", argv[1]);
		return 1;
	}

	vector<Instruction> program;
	Instruction instr;
	while (input >> instr)
	{
		program.emplace_back(move(instr));
	}
	input.close();

	Computer c;
	fmt::print("Part1: {}\n", c.execute(program));
	c.reset();
	c.registers['c'-'a'] = 1;
	fmt::print("Part2: {}\n", c.execute(program));

	return 0;
}

File: day12/day12.py
#!/usr/bin/env python3

import sys

class Instruction(object):
    def __init__(self, txt):
        lst = txt.split(" ")
        self.op = lst[0]
        if self.op in ("inc", "dec"):
            self.arg1, self.arg2 = lst[1], None
        else:
            self.arg1, self.arg2  = lst[1], lst[2]

class Computer(object):
    def __init__(self):
        self.reset()

    def resolve(self, arg):
        value = self.register.get(arg)
        if value is None:
            return int(arg)
        return value

    def reset(self):
        self.register = {v: 0 for v in "abcd"}

    def execute(self, program):
        ip = 0
        while 0 <= ip < len(program):
            i = program[ip]
            if i.op == "cpy":
                self.register[i.arg2] = self.resolve(i.arg1)
                ip += 1
            elif i.op == "inc":
                self.register[i.arg1] += 1
                ip += 1
            elif i.op == "dec":
                self.register[i.arg1] -= 1
                ip += 1
            elif i.op == "jnz":
                if self.resolve(i.arg1):
                    ip += self.resolve(i.arg2)
                else:
                    ip += 1
            else:
                break

        return self.resolve("a")

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read()
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

    program = [Instruction(line) for line in txt.splitlines()]
    c = Computer()
    print("Part1:", c.execute(program))
    c.reset()
    c.register["c"] = 1
    print("Part2:", c.execute(program))
